phoenix-core/src/main/java/org/apache/phoenix/cache/TenantCacheImpl.java

```diff
@@ -77,57 +152,132 @@ public MemoryManager getMemoryManager() {
+        if (isPersistent) {
+            builder.expireAfterWrite(ttl, TimeUnit.MILLISECONDS);
+        } else {
+            builder.expireAfterAccess(ttl, TimeUnit.MILLISECONDS);
+        }
+        return builder
+            .ticker(getTicker())
+            .removalListener(new RemovalListener<ImmutableBytesPtr, CacheEntry>(){
+                @Override
+                public void onRemoval(RemovalNotification<ImmutableBytesPtr, CacheEntry> notification) {
+                	if (isPersistent || !notification.getValue().getUsePersistentCache()) {
+                        Closeables.closeAllQuietly(Collections.singletonList(notification.getValue()));
+                	}
+                }
+            })
+            .build();
+    }
     
-    @Override
+    private void evictInactiveEntries(long bytesNeeded) {
+        CacheEntry[] entries = getPersistentServerCaches().asMap().values().toArray(new CacheEntry[]{});
+        Arrays.sort(entries);
+        long available = this.getMemoryManager().getAvailableMemory();
+        for (int i = 0; i < entries.length && available < bytesNeeded; i++) {
+            CacheEntry entry = entries[i];
+            if (!entry.isLive()) {
+            	getServerCaches().invalidate(entry.getCacheId());
+            	getPersistentServerCaches().invalidate(entry.getCacheId());
+                available = this.getMemoryManager().getAvailableMemory();
+            }
+        }
+    }
+
+    private CacheEntry getIfPresent(ImmutableBytesPtr cacheId) {
+        CacheEntry entry = getPersistentServerCaches().getIfPresent(cacheId);
+        if (entry != null) {
+            getServerCaches().put(cacheId, entry);
+        }return getServerCaches().getIfPresent(cacheId);
+    }
+
+	@Override
     public Closeable getServerCache(ImmutableBytesPtr cacheId) {
         getServerCaches().cleanUp();
-        return getServerCaches().getIfPresent(cacheId);
+        CacheEntry entry = getIfPresent(cacheId);
+        if (entry == null) {
+            return null;
+        }
+        return entry.closeable;
     }
-    
+
     @Override
-    public Closeable addServerCache(ImmutableBytesPtr cacheId, ImmutableBytesWritable cachePtr, byte[] txState, ServerCacheFactory cacheFactory, boolean useProtoForIndexMaintainer, int clientVersion) throws SQLException {
+    public Closeable addServerCache(ImmutableBytesPtr cacheId, ImmutableBytesWritable cachePtr, byte[] txState, ServerCacheFactory cacheFactory, boolean useProtoForIndexMaintainer, boolean usePersistentCache, int clientVersion) throws SQLException {
         getServerCaches().cleanUp();
-        MemoryChunk chunk = this.getMemoryManager().allocate(cachePtr.getLength() + txState.length);
+        long available = this.getMemoryManager().getAvailableMemory();
+        int size = cachePtr.getLength() + txState.length;
+        if (size > available) {
+            evictInactiveEntries(Math.max(size - available + EVICTION_MARGIN_BYTES, EVICTION_MARGIN_BYTES));
```

Comment:
Seems like this will always be size - available + EVICTION_MARGIN_BYTES since size>available. So if we need to evict we will evict a minimum of EVICTION_MARGIN_BYTES ?

Reply:
Good point, in fact the check is right above :). Removed the redundant `max`